src/main/scala/org/broadinstitute/hail/sparkextras/AdjustedPartitionsRDD.scala

```diff
@@ -0,0 +1,66 @@
+package org.broadinstitute.hail.sparkextras
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{Dependency, NarrowDependency, Partition, TaskContext}
+
+import scala.reflect.ClassTag
+
+case class AdjustedPartitionsRDDPartition[T](index: Int, parent: Partition, adj: Adjustment[T, Partition]) extends Partition
+
+case class Adjustment[T, U](originalIndex: Int, dropWhile: Option[T => Boolean], takeWhile: Option[(Seq[U], T => Boolean)]) {
+  def map[V](f: U => V): Adjustment[T, V] = copy(takeWhile = takeWhile.map { case (it, g) => (it.map(f), g) })
+}
+
+class AdjustedPartitionsRDD[T](@transient var prev: RDD[T], adjustInfo: IndexedSeq[Adjustment[T, Int]])(implicit tct: ClassTag[T])
+  extends RDD[T](prev.sparkContext, Nil) {
+  require(adjustInfo.forall(adj => adj.takeWhile.forall { case (it, _) =>
+    it.nonEmpty && it.head == adj.originalIndex + 1 && it.zip(it.tail).forall { case (l, r) => l + 1 == r }
+  }))
```

Comment:
I went back over this and it's actually NOT required that `adjustments.length == prev.partitions.length`.  It's possible for a locus stretch to span a partition (the `min == max` thing we talked about earlier) in which case that partition wouldn't have an adjustment.